### PR TITLE
Fix TypeScript typings

### PR DESCRIPTION
**Pull request by wipfli**
## Launch Checklist

Issue #776

 - [x] Confirm **your changes do not include backports from Mapbox projects** (unless with compliant license) - if you are not sure about this, please ask!
 - [ ] Briefly describe the changes in this PR.
 - [ ] Include before/after visuals or gifs if this PR includes visual changes.
 - [ ] Write tests for all new functionality.
 - [ ] Document any changes to public APIs.
 - [ ] Post benchmark scores.
 - [ ] Manually test the debug page.
 - [ ] Suggest a changelog category: bug/feature/docs/etc. or "skip changelog".
 - [ ] Add an entry inside this element for inclusion in the `maplibre-gl-js` changelog: `<changelog></changelog>`.

### CODE REVIEW

**Comment by HarelM** [comment]
I think I fixed most of the stuff here.
Not sure how to proceed.
I'll sleep on it...

**Comment by HarelM** [comment]
I had an idea this morning :-)
We are post-processing the d.ts file already, this is done to export classes as public due to UMD stuff, so we already have this in place.
What I thought about doing is adding the package-types files content to the begging of the generated file thus achieving the ability to declare the missing modules without the need to go through definitely typed or mapbox modules.
That can be one approach.
For vector-tile-js repo I don't think it should be incorporated in this repo, but the grid-index is a single file that I think we might be able to simply add it here.
We already have a grid-index in this repo which looks very similar. Adding the other grid index here might allow someone to take a look at both of them and find common ground, or not, which is fine too.
@astridx @wipfli if one of you guys would like to incorporate grid-index package here it might simplify things a bit for this PR.
I'll work later tonight on updating the post build script to incorporate the index.d.ts files in types-packages folder. the less files we have there the better we'll be :-)

**Comment by wipfli** [comment]
I'm happy to look into this later tonight.

**Comment by wipfli** [comment]
> Confirm your changes do not include backports from Mapbox projects (unless with compliant license) - if you are not sure about this, please ask!

I copied `grid-index.js` from Mapbox https://github.com/mapbox/grid-index and migrated it to TypeScript (https://github.com/maplibre/maplibre-gl-js/pull/787/commits/4a443dba3513540d4bb392fe37fc9e41eea0fed3). The license of the copied source code is 
```
Copyright (c) 2016, Mapbox

Permission to use, copy, modify, and/or distribute this software for any purpose
with or without fee is hereby granted, provided that the above copyright notice
and this permission notice appear in all copies.

THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY AND
FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM LOSS
OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER
TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR PERFORMANCE OF
THIS SOFTWARE.
``` 
I included a copy of the license at the beginning of the new file.

Let me know if I need to to something with regards to licensing.

**Comment by HarelM** [comment]
@wipfli thanks for the import! Do you mind copying the tests as well?
I think that the generated d.ts file now is OK and can be used.
The only error is get is that you need to install `@types/pbf` in order to use maplibre, I'll see if I can find some info on how to do it "automatically" when one installs maplibre.

**Comment by HarelM** [comment]
According to this SO answer:
https://stackoverflow.com/a/46011417/2946662
I need to move @types/geojson and @type/pbf to dependencies and remove them from devDependencies.
In order to properly solve this issue I think.
I'm, not sure how to test the dependencies part without publishing a package.
In any case, I think this PR is ready for review...?

File: build/post-ts-build.js [hunk 1]
```diff
@@ -35,7 +35,17 @@ console.log(`Starting bundling types`);
 const outputFile = "./dist/maplibre-gl.d.ts";
 child_process.execSync(`dts-bundle-generator --no-check --umd-module-name=maplibregl -o ${outputFile} ./src/index.ts`);
```
Comment (wipfli) [inline build/post-ts-build.js:36]: `--no-check` still needed?

Comment (HarelM) [reply build/post-ts-build.js:36]: Yes, unfortunately...

File: src/types/packages-types/vector-tile/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,27 @@
+declare module '@mapbox/vector-tile' {
```
Comment (wipfli) [inline src/types/packages-types/vector-tile/index.d.ts:1]: We don't need a license remark for types only, right?

Comment (HarelM) [reply src/types/packages-types/vector-tile/index.d.ts:1]: Notv that I know of, no.

File: src/util/transferable_grid_index.test.ts [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+    test('returns multiple copies of a key if multiple boxes were inserted with the same key', () => {
+        const grid = new TransferableGridIndex(100, 4, 0);
+        const key = 123;
+        grid.insert(key, 3, 3, 4, 4);
+        grid.insert(key, 13, 13, 14, 14);
+        grid.insert(key, 23, 23, 24, 24);
+        expect(grid.query(0, 0, 30, 30)).toEqual([key, key, key]);
+    });
+
+    test('serializing to an arraybuffer', () => {
```
Comment (wipfli) [inline src/util/transferable_grid_index.test.ts:27]: Probably would be good to add a unit test for serialize and deserialize too

**Comment by wipfli** [comment]
I reviewed all changes and they look good to me. Basically everything was written by @HarelM, I only opened this pull request from a branch that Harel created and migrated the `grind-index` file from Mapbox.

@HarelM, since I opened this pull request, I cannot approve it, so please approve it yourself.

File: src/util/transferable_grid_index.test.ts (hunk 3 above)
Comment (HarelM) [reply src/util/transferable_grid_index.test.ts:27]: Added. thanks for the review!

**Comment by HarelM** [comment]
Should I move @types/pbf to dev dependencies? It is needed for people using this package so that the typings will be fully OK...

**Comment by wipfli** [comment]
Please add them the dependencies, yes.

**Review by HarelM — APPROVED** [review]
Approved by @wipfli and myself

**Comment by wipfli** [comment]
Let's hope the build-release-notes script gets the text right this time...

**Comment by wipfli** [comment]
Nope, it is again off by one...

**Comment by wipfli** [comment]
#780

**Pull request merged**